### PR TITLE
Change stop colors for routes

### PR DESCRIPTION
**Pull request by tliss**
This is for WMATA-309.

### CODE REVIEW

**Comment by tliss** [comment]
This is one of four pull requests for WMATA-309. The requests are:

https://github.com/camsys/onebusaway-application-modules/pull/16
https://github.com/camsys/onebusaway-hart-enterprise/pull/2
https://github.com/camsys/onebusaway-wmata-enterprise/pull/1
https://github.com/camsys/onebusaway-sound-enterprise/pull/4

File: onebusaway-enterprise-webapp/src/main/java/org/onebusaway/enterprise/webapp/actions/api/StopsOnRouteAction.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    for(StopBean stopBean : stopsForRoute.getStops()) {
+      stopIdToStopBeanMap.put(stopBean.getId(), stopBean);
+    }
+    
+    // break up stops into destinations
+    List<StopGroupingBean> stopGroupings = stopsForRoute.getStopGroupings();
+    for (StopGroupingBean stopGroupingBean : stopGroupings) {
+      for (StopGroupBean stopGroupBean : stopGroupingBean.getStopGroups()) {
+        NameBean name = stopGroupBean.getName();
+        String type = name.getType();
```
Comment (sheldonabrown) [inline onebusaway-enterprise-webapp/src/main/java/org/onebusaway/enterprise/webapp/actions/api/StopsOnRouteAction.java]: Why is this here?

File: onebusaway-enterprise-webapp/src/main/java/org/onebusaway/enterprise/webapp/actions/api/StopsOnRouteAction.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        
+        if(!stopGroupBean.getStopIds().isEmpty()) {
+          for(String stopId : stopGroupBean.getStopIds()) {
+            String agencyId = AgencyAndIdLibrary.convertFromString(_routeId).getAgencyId();
```
Comment (sheldonabrown) [inline onebusaway-enterprise-webapp/src/main/java/org/onebusaway/enterprise/webapp/actions/api/StopsOnRouteAction.java]: Careful!  The agency of the stop can and will be different from the route.  You want stopBean.getId() here

File: onebusaway-enterprise-webapp/src/main/java/org/onebusaway/enterprise/webapp/actions/api/StopsOnRouteAction.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+@ParentPackage("json-default")
+@Result(type="json", params={"callbackParameter", "callback"})
+public class StopsOnRouteAction extends OneBusAwayEnterpriseActionSupport {
+
+  private static final long serialVersionUID = 1L;
+
+  @Autowired
+  private TransitDataService _transitDataService;
+
+  private List<StopOnRoute> _stops = new ArrayList<StopOnRoute>();
```
Comment (sheldonabrown) [inline onebusaway-enterprise-webapp/src/main/java/org/onebusaway/enterprise/webapp/actions/api/StopsOnRouteAction.java]: I think this wants to be a Set, not a List, as you may be iterating over multiple stop groupings and therefor may have duplicate stops

File: onebusaway-enterprise-webapp/src/main/java/org/onebusaway/enterprise/webapp/actions/api/StopsOnRouteAction.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
+/**
+ * Copyright (C) 2011 Metropolitan Transportation Authority
```
Comment (sheldonabrown) [inline onebusaway-enterprise-webapp/src/main/java/org/onebusaway/enterprise/webapp/actions/api/StopsOnRouteAction.java]: This should be Copyright (C) 2018 Cambridge Systematics, Inc.

File: onebusaway-enterprise-webapp/src/main/java/org/onebusaway/enterprise/webapp/actions/api/StopsOnRouteAction.java (hunk 1 above)
Comment (sheldonabrown) [reply onebusaway-enterprise-webapp/src/main/java/org/onebusaway/enterprise/webapp/actions/api/StopsOnRouteAction.java]: I understand now.  You copy and pasted from StopsOnRouteForDirectionAction.  Please don't do that.  Read https://en.wikipedia.org/wiki/Don%27t_repeat_yourself

It seems like by adding an additional flag to stops-on-route-for-direction or optionally letting directionId be null you could simply re-use this API call.

File: onebusaway-enterprise-webapp/src/main/java/org/onebusaway/enterprise/webapp/actions/api/StopsOnRouteAction.java (hunk 2 above)
Comment (sheldonabrown) [reply onebusaway-enterprise-webapp/src/main/java/org/onebusaway/enterprise/webapp/actions/api/StopsOnRouteAction.java]: I now see this isn't your fault.  Its still wrong tho :)

File: onebusaway-frontend-webapp/src/main/webapp/js/oba/map/Sidebar.js [hunk 5]
```diff
@@ -275,16 +275,29 @@ OBA.Sidebar = function() {
...
+		// Need stop info for highlighting stops in RouteMap.js
+		var savedData = { stops:[] };
+
 		jQuery.each(routeResults, function(_, routeResult) {
 			if (routeResult.shortName === filter) {
 				filterExistsInResults = true;
 				return false;
 			}
+
+            jQuery.getJSON(OBA.Config.stopsOnRoute + "?callback=?", { routeId: routeResult.id }, function (json) {
```
[the remark this answers is not in the document]
Comment (sheldonabrown) [reply onebusaway-frontend-webapp/src/main/webapp/js/oba/map/Sidebar.js]: Thinking more on this, with a little refactoring you could load the stop data (as savedData) after a search and have both loadStopsForRouteAndDirection and addRoutesToLegend use it.  Rebecca can help you figure this out....Just remember that based on the refactoring Rebecca is doing the SideBar wont always be loaded (for the mobile view)

**Comment by tliss** [comment]
Closing as these changes are no longer needed.

File: onebusaway-enterprise-webapp/src/main/java/org/onebusaway/enterprise/webapp/actions/api/model/StopResult.java [hunk 6]
```diff
@@ -32,17 +34,39 @@
...
+	private List<StopBean> stops;
+
+	private List<String> stopIdsForRoute = new ArrayList<>();
+
 	public StopResult(StopBean stop, List<RouteAtStop> routesAvailable) {
 		this.stop = stop;
 		this.routesAvailable = routesAvailable;
 	}
+
+	public StopResult(StopBean stop, List<RouteAtStop> routesAvailable, List<StopsForRouteBean> fullStopList) {
```
Comment (tliss) [inline onebusaway-enterprise-webapp/src/main/java/org/onebusaway/enterprise/webapp/actions/api/model/StopResult.java]: I added a new signature, rather than changing the old one, because I wasn't sure if the old one was still being used in other locations.

File: onebusaway-enterprise-webapp/src/main/java/org/onebusaway/enterprise/webapp/actions/api/model/RouteResult.java [hunk 7]
```diff
@@ -32,11 +35,28 @@
...
+	private List<StopBean> stops;
+
+	private List<String> stopIdsForRoute = new ArrayList<>();
+
 	public RouteResult(RouteBean route, List<RouteDirection> directions) {
 		this.route = route;
 		this.directions = directions;
 	}
 
+	public RouteResult(RouteBean route, List<RouteDirection> directions, StopsForRouteBean stopsForRouteBean) {
```
Comment (tliss) [inline onebusaway-enterprise-webapp/src/main/java/org/onebusaway/enterprise/webapp/actions/api/model/RouteResult.java]: I added a new signature, rather than changing the old one, because I wasn't sure if the old one was still being used in other locations.

File: onebusaway-enterprise-webapp/src/main/java/org/onebusaway/enterprise/webapp/actions/api/SearchResultFactoryImpl.java [hunk 8]
```diff
@@ -55,9 +55,9 @@ public SearchResultFactoryImpl(SearchService searchService, TransitDataService t
   }
 
   @Override
-  public SearchResult getRouteResultForRegion(RouteBean routeBean) { 
+  public SearchResult getRouteResultForRegion(RouteBean routeBean) {
     List<String> polylines = new ArrayList<String>();
-
```
Comment (sheldonabrown) [inline onebusaway-enterprise-webapp/src/main/java/org/onebusaway/enterprise/webapp/actions/api/SearchResultFactoryImpl.java]: If at all possible, try to clean up pull requests such that white space is not significant.  Its a nit, but across a larger pull request it makes it really hard to tell what was changed.  No need to make changes to this one -- but something to consider in the future

File: onebusaway-enterprise-webapp/src/main/java/org/onebusaway/enterprise/webapp/actions/api/model/RouteResult.java (hunk 7 above)
Comment (sheldonabrown) [reply onebusaway-enterprise-webapp/src/main/java/org/onebusaway/enterprise/webapp/actions/api/model/RouteResult.java]: That's fine.  Looks good
